flocker/common/test/test_algebraic.py

```diff
@@ -0,0 +1,150 @@
+# -*- test-case-name: flocker.common.test.test_algebraic -*-
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+
+"""
+Tests for ``flocker.common.algebraic``
+"""
+
+from pyrsistent import (
+    PClass, field,
+    InvariantException,
+)
+from hypothesis import given, strategies as st, assume
+
+from twisted.trial.unittest import SynchronousTestCase
+
+from twisted.python.constants import Names, NamedConstant
+
+
+from ..algebraic import TaggedUnionInvariant, tagged_union_strategy
+
+
+class States(Names):
+    ALLOWED = NamedConstant()
+    WITH_ATTRIBUTE = NamedConstant()
+    WITH_TWO_ATTRIBUTES = NamedConstant()
+    DISALLOWED = NamedConstant()
+
+
+class AlgebraicType(PClass):
+    state = field(mandatory=True)
+    one = field(bool)
+    two = field(bool)
+    extra = field(bool, mandatory=True)
+
+    __invariant__ = TaggedUnionInvariant(
+        tag_attribute='state',
+        attributes_for_tag={
+            States.ALLOWED: set(),
+            States.WITH_ATTRIBUTE: {'one'},
+            States.WITH_TWO_ATTRIBUTES: {'one', 'two'},
+        },
+    )
+
+ALGEBRAIC_TYPE_STRATEGY = tagged_union_strategy(AlgebraicType, {
+    'one': st.booleans(),
+    'two': st.booleans(),
+    'extra': st.booleans(),
+})
+
+ALGEBRAIC_TYPE_ARGUMENTS_STRATGEY = ALGEBRAIC_TYPE_STRATEGY.map(
+    lambda v: v.serialize())
+
+
+class TaggedUnionInvariantTests(SynchronousTestCase):
+    """
+    Tests for ``TaggedUnionInvariant``.
+    """
+
+    @given(
+        args=ALGEBRAIC_TYPE_ARGUMENTS_STRATGEY,
+    )
+    def test_valid_strategy(self, args):
+        """
+        When a valid dictionary of attributes is provided, an
+        instance of ``AlgebraicType`` is provided.
+        """
+        self.assertIsInstance(AlgebraicType(**args), AlgebraicType)
+
+    @given(
+        args=ALGEBRAIC_TYPE_ARGUMENTS_STRATGEY,
+        choice=st.choices(),
+        extra_value=st.booleans()
+    )
+    def test_extra_attributes(self, args, choice, extra_value):
+        """
+        When an extra attribute that isn't allowed in a given state
+        is provided, ``InvariantException`` is raised.
+
+        :param choice: A choice function
+        :param extra_value: A value to provide to the exta attribute
+        """
+        state = args['state']
+        invariant = AlgebraicType.__invariant__
+        extra_attributes = (
+            invariant._all_attributes
+            - invariant.attributes_for_tag[state]
+        )
+        assume(extra_attributes)
+        extra_attribute = choice(sorted(extra_attributes))
```

Comment:
Yes. `choice` doesn't accept `set`s and since hypothesis keeps examples around, I thought it would be good to have a stable ordering to pick from.